[bug] "IMPORTED_IMPLIB not set for imported target" when consuming mimalloc as shared through CMakeConfigDeps

### Describe the bug

I'm not sure if I should post this here on on the CCI, but since it's only happening with the new `CMakeConfigDeps` generator, I opted to post it here!

When consuming `mimalloc/2.2.4` as a **shared** library on Windows through `CMakeConfigDeps`,
I get an error from CMake:
```
(conan2-venv) D:\Dev\drone_flight_planner\bug_repro>cmake --preset conan-default
-- Using Conan toolchain: D:/Dev/drone_flight_planner/bug_repro/generators/conan_toolchain.cmake
-- Conan toolchain: CMAKE_GENERATOR_TOOLSET=v143
-- Conan toolchain: Setting CMAKE_MSVC_RUNTIME_LIBRARY=$<$<CONFIG:Debug>:MultiThreadedDebugDLL>$<$<CONFIG:Release>:MultiThreadedDLL>
-- Conan toolchain: C++ Standard 14 with extensions OFF
-- Conan toolchain: Including CMakeConfigDeps generated conan_cmakedeps_paths.cmake
-- Selecting Windows SDK version 10.0.26100.0 to target Windows 10.0.22631.
-- The CXX compiler identification is MSVC 19.44.35227.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/18/Professional/VC/Tools/MSVC/14.44.35207/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Configuring Targets for mimalloc/2.2.4
-- Conan: Target declared imported SHARED library 'mimalloc'
-- Configuring done (3.5s)
CMake Error in CMakeLists.txt:
  IMPORTED_IMPLIB not set for imported target "mimalloc" configuration
  "Debug".


CMake Error in CMakeLists.txt:
  IMPORTED_IMPLIB not set for imported target "mimalloc" configuration
  "Release".


CMake Error in CMakeLists.txt:
  IMPORTED_IMPLIB not set for imported target "mimalloc" configuration
  "MinSizeRel".


CMake Error in CMakeLists.txt:
  IMPORTED_IMPLIB not set for imported target "mimalloc" configuration
  "RelWithDebInfo".
```

The issue doesn't reproduce if I consume `mimalloc` as a **static** library, or if I use `CMakeDeps` instead of `CMakeConfigDeps`.

### How to reproduce it

## Environment

| Field | Value |
|---|---|
| Conan version | 2.29.0 |
| OS | Windows 11 |
| Compiler | MSVC 19.44 |
| CMake | ≥ 3.25 (I'm using 4.2.3) |
| Generator | `CMakeConfigDeps` (experimental) |

## Conan profile

```
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.version=194
os=Windows
```

---

## Steps to Reproduce

### `conanfile.py`

```python
from conan import ConanFile
from conan.tools.cmake import CMakeToolchain, CMakeConfigDeps, cmake_layout

class BugReproConsumer(ConanFile):
    name = "bug_repro"
    version = "0.1"
    settings = "os", "arch", "compiler", "build_type"

    def requirements(self):
        self.requires("mimalloc/2.2.4")

    def layout(self):
        cmake_layout(self)

    def generate(self):
        CMakeConfigDeps(self).generate()
        CMakeToolchain(self).generate()
```

### `CMakeLists.txt`

```cmake
cmake_minimum_required(VERSION 3.25)
project(bug_repro CXX)

find_package(mimalloc CONFIG REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE mimalloc)
```

### `main.cpp`

```cpp
#include <mimalloc.h>
int main() {
    void* p = mi_malloc(64);
    mi_free(p);
    return 0;
}
```

### Commands

```bat
conan install . -o *:shared=True

cmake --preset conan-default
```

--- a/conan/internal/model/cpp_info.py
+++ b/conan/internal/model/cpp_info.py
@@ -577,36 +577,25 @@ class _Component:
         libdirs = self.libdirs
         bindirs = self.bindirs
         libname = self.libs[0]
-        static_location = None
-        shared_location = None
         dll_location = None
         deduced_type = None
-        # libname is exactly the pattern, e.g., ["mylib.a"] instead of ["mylib"]
-        _, ext = os.path.splitext(libname)
-        if ext in (".lib", ".a", ".dll", ".so", ".dylib"):
-            if ext in (".lib", ".a"):
-                static_location = _find_matching(libdirs, libname)
-            elif ext in (".so", ".dylib"):
-                shared_location = _find_matching(libdirs, libname)
-            elif ext == ".dll":
-                dll_location = _find_matching(bindirs, libname)
-        else:
-            lib_sanitized = re.escape(libname)
-            component_sanitized = re.escape(library_name)
-            # At first, exact match
-            regex_static = re.compile(rf"(?:lib)?{lib_sanitized}\.(?:a|lib)")
-            regex_shared = re.compile(rf"(?:lib)?{lib_sanitized}\.(?:so|dylib)")
-            regex_dll = re.compile(rf".*(?:{lib_sanitized}|{component_sanitized}).*\.dll")
-            static_location = _find_matching(libdirs, regex_static)
-            shared_location = _find_matching(libdirs, regex_shared)
-            if not any([static_location, shared_location]):
-                # Let's extend a little bit the pattern search
-                regex_wider_static = re.compile(rf"(?:lib)?{lib_sanitized}(?:[._-].+)?\.(?:a|lib)")
-                regex_wider_shared = re.compile(rf"(?:lib)?{lib_sanitized}(?:[._-].+)?\.(?:so|dylib)")
-                static_location = _find_matching(libdirs, regex_wider_static)
-                shared_location = _find_matching(libdirs, regex_wider_shared)
-            if static_location or not shared_location:
-                dll_location = _find_matching(bindirs, regex_dll)
+        libname, ext = os.path.splitext(libname)
+        lib_sanitized = re.escape(libname)
+        component_sanitized = re.escape(library_name)
+        # At first, exact match
+        regex_static = re.compile(rf"(?:lib)?{lib_sanitized}\.(?:a|lib)")
+        regex_shared = re.compile(rf"(?:lib)?{lib_sanitized}\.(?:so|dylib)")
+        regex_dll = re.compile(rf".*(?:{lib_sanitized}|{component_sanitized}).*\.dll")
+        static_location = _find_matching(libdirs, regex_static)
+        shared_location = _find_matching(libdirs, regex_shared)
+        if not any([static_location, shared_location]):
+            # Let's extend a little bit the pattern search
+            regex_wider_static = re.compile(rf"(?:lib)?{lib_sanitized}(?:[._-].+)?\.(?:a|lib)")
+            regex_wider_shared = re.compile(rf"(?:lib)?{lib_sanitized}(?:[._-].+)?\.(?:so|dylib)")
+            static_location = _find_matching(libdirs, regex_wider_static)
+            shared_location = _find_matching(libdirs, regex_wider_shared)
+        if static_location or not shared_location:
+            dll_location = _find_matching(bindirs, regex_dll)
 
         if static_location:
             if shared_location:

--- a/test/integration/toolchains/cmake/cmakeconfigdeps/test_cmakeconfigdeps.py
+++ b/test/integration/toolchains/cmake/cmakeconfigdeps/test_cmakeconfigdeps.py
@@ -1031,3 +1031,26 @@ def test_libs_no_components_multilib_component(requires):
     tc.run("install --requires=matrix/1.0 -g CMakeConfigDeps")
     matrix_targets = tc.load("matrix-Targets-release.cmake")
     assert "# Requirement matrix::_common -> base::base (Full link: True)" in matrix_targets
+
+
+def test_implib_location_explicit_extension():
+    """
+    https://github.com/conan-io/conan/issues/20054
+    If the extension was explicitly added, we would not find the importlib if it was named
+    .dll.lib
+    """
+    tc = TestClient()
+    conanfile = (GenConanfile("allocator", "1.0")
+                 .with_package_file("lib/allocator.dll.lib", "importlib")
+                 .with_package_file("bin/allocator.dll", "dll")
+                 .with_package_type("shared-library")
+                 .with_package_info({"libs": ["allocator.dll"]}))
+
+    tc.save({"conanfile.py": conanfile})
+    tc.run("create")
+    tc.run("install --requires=allocator/1.0 -g=CMakeConfigDeps")
+    targets = tc.load("allocator-Targets-release.cmake")
+    # We find the importlib
+    assert "PROPERTIES IMPORTED_IMPLIB_RELEASE" in targets
+    # And we find the dll
+    assert "PROPERTIES IMPORTED_LOCATION_RELEASE" in targets

--- a/test/unittests/model/build_info/test_deduce_locations.py
+++ b/test/unittests/model/build_info/test_deduce_locations.py
@@ -97,7 +97,9 @@ def test_complex_deduce_locations_shared(lib_name, libs, conanfile):
     ("libcurl_imp.lib", "libcurl.dll", ["libcurl_imp"], "libcurl"),
     ("libcrypto.lib", "libcrypto-3-x64.dll", ["libcrypto"], "crypto"),
     ("libssl.lib", "libssl-3-x64.dll", ["libssl"], "ssl"),
-    ("zdll.lib", "zlib1.dll", ["zdll"], "zlib")
+    ("zdll.lib", "zlib1.dll", ["zdll"], "zlib"),
+    ("mimalloc.dll.lib", "mimalloc.dll", ["mimalloc"], "mimalloc"),
+    ("mimalloc.dll.lib", "mimalloc.dll", ["mimalloc.dll"], "mimalloc"),
 ])
 def test_windows_shared_link_locations(lib_name, dll_name, libs, pkg_name, conanfile):
     """

